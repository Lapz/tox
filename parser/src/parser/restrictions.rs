#[derive(Debug, Default, Clone, Copy)]
pub(crate) struct Restrictions {
    pub forbid_record: bool,
}

impl Restrictions {
    pub fn no_records() -> Self {
        Self {
            forbid_record: true,
        }
    }
}
