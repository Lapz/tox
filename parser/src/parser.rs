mod classes;
mod enums;
mod expressions;
mod function;
mod imports;
mod module;
mod params;
mod pattern;
mod pratt;
mod restrictions;
mod source_file;
mod type_alias;
mod type_args;
mod type_params;
mod types;
mod visibility;
use crate::Span;
use errors::{pos::Position, Reporter};
use pratt::{InfixParser, Precedence, PrefixParser, Rule as _, RuleToken};
use restrictions::Restrictions;
use rowan::{GreenNodeBuilder, TextRange, TextUnit};
use std::collections::HashMap;
use std::mem::replace;
use syntax::{
    SyntaxKind::{self, *},
    Token, T,
};

#[derive(Debug)]
enum State {
    PendingStart,
    Normal,
    PendingFinish,
}

pub struct Parser<'a> {
    input: &'a str,
    current: &'a Span<Token>,
    state: State,
    pub builder: GreenNodeBuilder<'static>,
    reporter: Reporter,
    text_pos: TextUnit,
    tokens: &'a [Span<Token>],
    token_pos: usize,
    prefix: HashMap<RuleToken, &'a dyn PrefixParser>,
    infix: HashMap<RuleToken, &'a dyn InfixParser>,
}

impl<'a> Parser<'a> {
    pub fn new(tokens: &'a [Span<Token>], reporter: Reporter, input: &'a str) -> Self {
        let mut parser = Parser {
            current: &tokens[0],
            tokens,
            state: State::PendingStart,
            token_pos: 0,
            text_pos: 0.into(),
            builder: GreenNodeBuilder::new(),
            prefix: HashMap::new(),
            infix: HashMap::new(),
            reporter,
            input,
        };
        parser.prefix(RuleToken::LBracket, &expressions::LiteralParselet);
        parser.prefix(RuleToken::Literal, &expressions::LiteralParselet);
        parser.prefix(RuleToken::Ident, &expressions::IdentParselet);
        parser.prefix(RuleToken::Excl, &expressions::UnaryParselet);
        parser.prefix(RuleToken::Minus, &expressions::UnaryParselet);
        parser.prefix(RuleToken::LParen, &expressions::GroupingParselet);
        parser.prefix(RuleToken::Pipe, &expressions::ClosureParselet);

        parser.infix(
            RuleToken::LBrace,
            &expressions::RecordParselet(Precedence::Call),
        );

        parser.infix(
            RuleToken::LParen,
            &expressions::CallParselet(Precedence::Call),
        );

        parser.infix(
            RuleToken::LBracket,
            &expressions::IndexParselet(Precedence::Call),
        );

        parser.infix(
            RuleToken::Dot,
            &expressions::FieldParselet(Precedence::Call),
        );
        parser.infix(
            RuleToken::Eq,
            &expressions::BinaryParselet(Precedence::Assignment),
        );

        parser.infix(
            RuleToken::PlusEq,
            &expressions::BinaryParselet(Precedence::Assignment),
        );

        parser.infix(
            RuleToken::StarEq,
            &expressions::BinaryParselet(Precedence::Assignment),
        );

        parser.infix(
            RuleToken::MinusEq,
            &expressions::BinaryParselet(Precedence::Assignment),
        );

        parser.infix(
            RuleToken::SlashEq,
            &expressions::BinaryParselet(Precedence::Assignment),
        );

        parser.infix(
            RuleToken::Plus,
            &expressions::BinaryParselet(Precedence::Term),
        );
        parser.infix(
            RuleToken::Minus,
            &expressions::BinaryParselet(Precedence::Term),
        );
        parser.infix(
            RuleToken::Slash,
            &expressions::BinaryParselet(Precedence::Term),
        );
        parser.infix(
            RuleToken::Star,
            &expressions::BinaryParselet(Precedence::Term),
        );
        parser.infix(
            RuleToken::Comparison,
            &expressions::BinaryParselet(Precedence::Comparison),
        );

        parser.infix(
            RuleToken::AmpAmp,
            &expressions::BinaryParselet(Precedence::And),
        );

        parser.infix(
            RuleToken::PipePipe,
            &expressions::BinaryParselet(Precedence::Or),
        );

        parser.infix(
            RuleToken::EqEq,
            &expressions::BinaryParselet(Precedence::Equality),
        );

        parser
    }

    pub fn reporter(self) -> Reporter {
        self.reporter
    }

    fn prefix(&mut self, rule: RuleToken, parser: &'a dyn pratt::PrefixParser) {
        self.prefix.insert(rule, parser);
    }

    fn infix(&mut self, rule: RuleToken, parser: &'a dyn pratt::InfixParser) {
        self.infix.insert(rule, parser);
    }

    fn checkpoint(&self) -> rowan::Checkpoint {
        self.builder.checkpoint()
    }

    fn lookahead(&self) -> Option<&Span<Token>> {
        self.tokens.get(self.token_pos + 1)
    }

    fn start_node(&mut self, kind: SyntaxKind) {
        match replace(&mut self.state, State::Normal) {
            State::PendingStart => {
                self.builder.start_node(kind.into());
                // No need to attach trivia's to previous node: there is no
                // previous node.
                return;
            }
            State::PendingFinish => (),
            State::Normal => (),
        }

        let n_trivias = self.tokens[self.token_pos..]
            .iter()
            .take_while(|token| token.value.kind.is_trivia())
            .count();

        let leading_trivias = &self.tokens[self.token_pos..self.token_pos + n_trivias];
        let mut trivia_end: TextUnit = self.text_pos
            + leading_trivias
                .iter()
                .map(|it| TextUnit::from(it.value.len))
                .sum::<TextUnit>();

        let n_attached_trivias = {
            let leading_trivias = leading_trivias.iter().rev().map(|it| {
                let next_end = trivia_end - TextUnit::from(it.value.len);
                let range = TextRange::from_to(next_end, trivia_end);
                trivia_end = next_end;
                (it.value.kind, &self.input[range])
            });
            n_attached_trivias(kind, leading_trivias)
        };

        self.eat_n_trivias(n_trivias - n_attached_trivias);

        self.builder.start_node(kind.into());

        self.eat_n_trivias(n_attached_trivias)
    }

    fn eat_n_trivias(&mut self, n: usize) {
        for _ in 0..n {
            let token = &self.tokens[self.token_pos];
            assert!(token.value.kind.is_trivia());
            self.add_token(token);
        }
    }

    fn start_node_at(&mut self, checkpoint: rowan::Checkpoint, kind: SyntaxKind) {
        self.builder.start_node_at(checkpoint, kind.into());
    }

    fn finish_node(&mut self) {
        match replace(&mut self.state, State::PendingFinish) {
            State::PendingFinish | State::Normal => {
                self.builder.finish_node();
            }
            State::PendingStart => unreachable!(),
        }
    }

    fn recover(&mut self) {
        match self.current() {
            T!["{"] => self.recover_until(T!["}"]),
            T!["}"] => self.recover_until(T![;]),
            IDENT => self.recover_until_many(&[T!["}"], T![;], T![fn]]),
            _ => {
                self.bump();
            }
        }
    }

    fn recover_until_many(&mut self, tokens: &[SyntaxKind]) {
        while self.lookahead().is_some() && !self.matches(tokens.to_vec()) {
            self.bump();
        }

        self.bump(); // eat the token as well
    }

    fn recover_until(&mut self, token: SyntaxKind) {
        while self.lookahead().is_some() && !self.at(token) {
            self.bump();
        }

        self.bump(); // eat the token as well
    }

    fn error(&mut self, message: impl Into<String>, additional_info: impl Into<String>) {
        let message = message.into();
        let additional_info = additional_info.into();
        self.start_node(SyntaxKind::ERROR);

        self.recover();

        self.reporter
            .error(message, additional_info, self.current_span());
        self.finish_node()
    }

    fn error_until(
        &mut self,
        message: impl Into<String>,
        additional_info: impl Into<String>,
        token: SyntaxKind,
    ) {
        self.start_node(SyntaxKind::ERROR);

        self.reporter
            .error(message, additional_info, self.current_span());

        self.recover_until(token);

        self.finish_node()
    }

    fn current_span(&self) -> (Position, Position) {
        (self.current.start, self.current.end)
    }

    fn current_string(&self) -> &str {
        let token = self.current;

        &self.input[token.start.absolute as usize..token.end.absolute as usize]
    }

    fn precedence(&mut self) -> Precedence {
        let token = self.current();

        let rule = token.rule();

        self.infix
            .get(&rule)
            .map_or(Precedence::None, |parser| parser.pred())
    }

    fn expect(&mut self, expected: SyntaxKind) {
        if self.at(expected) {
            self.bump();
        } else {
            let current = self.current();
            self.error(
                format!("Expected `{}`", expected.text()),
                format!(
                    "Expected `{}` but instead found `{}`",
                    expected.text(),
                    current.text()
                ),
            );
        }
    }

    fn expects(&mut self, expected: Vec<SyntaxKind>) -> bool {
        for kind in expected {
            let current = self.current();
            if kind == current {
                self.bump();
                return true;
            }
        }

        false
    }

    fn expected(&mut self, expected: SyntaxKind) -> bool {
        if self.at(expected) {
            self.bump();
            true
        } else {
            self.error_until(
                format!("Expected `{}`", expected.text()),
                format!(
                    "Expected `{}` but instead found `{}` ",
                    expected.text(),
                    self.current_string()
                ),
                expected,
            );
            false
        }
    }

    fn current(&mut self) -> SyntaxKind {
        while self.current.value.kind.is_trivia() {
            self.add_token(self.current);
        }
        self.current.value.kind
    }

    fn at(&mut self, check: SyntaxKind) -> bool {
        self.current() == check
    }

    fn matches(&mut self, kind: Vec<SyntaxKind>) -> bool {
        for kind in kind {
            if kind == self.current() {
                return true;
            }
        }
        false
    }

    fn add_token(&mut self, token: &Span<Token>) {
        let len = TextUnit::from(token.value.len);
        let range = TextRange::offset_len(token.start.absolute.into(), len);

        let text = &self.input[range];

        self.text_pos += len;
        self.token_pos += 1;

        self.current = &self.tokens[self.token_pos];
        self.builder.token(token.value.kind.into(), text.into());
    }

    pub fn bump(&mut self) {
        if self.at(SyntaxKind::EOF) {
            return;
        }

        self.add_token(self.current);
    }

    fn ident(&mut self) {
        self.start_node(NAME);
        self.expect(IDENT);
        self.finish_node()
    }
}

/// Taken from https://github.com/rust-analyzer/rust-analyzer/blob/918547dbe9a2907401102eba491ac25cebe1404d/crates/ra_syntax/src/parsing/text_tree_sink.rs
/// All copyright goes to the rust-analyzer devs
fn n_attached_trivias<'a>(
    kind: SyntaxKind,
    trivias: impl Iterator<Item = (SyntaxKind, &'a str)>,
) -> usize {
    use SyntaxKind::*;
    match kind {
        TYPE_ALIAS_DEF | CLASS_DEF | ENUM_DEF | ENUM_VARIANT | FN_DEF | MOD_DEF | IMPORT_DEF => {
            let mut res = 0;
            for (i, (kind, text)) in trivias.enumerate() {
                match kind {
                    WHITESPACE => {
                        if text.contains("\n\n") {
                            break;
                        }
                    }
                    COMMENT => {
                        res = i + 1;
                    }
                    _ => (),
                }
            }
            res
        }
        _ => 0,
    }
}
