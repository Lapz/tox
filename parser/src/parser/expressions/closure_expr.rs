use syntax::T;

use crate::parser::Parser;

use crate::SyntaxKind::*;

use crate::parser::PrefixParser;

#[derive(Debug)]
pub struct ClosureParselet;

impl<'a> Parser<'a> {
    pub(crate) fn parse_closure_expr(&mut self) {
        self.start_node(CLOSURE_EXPR);

        self.parse_func_params(T![|]);

        if self.current() == T![->] {
            self.parse_return_type();
        }

        self.parse_block();

        self.finish_node()
    }
}

impl PrefixParser for ClosureParselet {
    fn parse(&self, parser: &mut Parser) {
        parser.parse_closure_expr();
    }
}

#[cfg(test)]
mod tests {
    test_parser! {parse_simple_closure_expr,r"
        fn main() {
            let a = |x,y| {
                return x+y;
            }
    }"}
    test_parser! {parse_closure_with_types,"fn main() {
        let a = |x,y| -> i32 {
            return x+y; 
        }
    }"}
    test_parser! {parse_empty_closure_expr,"fn main() {
        let a = || {};
    }"}
    test_parser! {parse_free_closure_expr,"fn main() {
        || {};
    }"}
}
