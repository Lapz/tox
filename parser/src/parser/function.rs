use syntax::T;

use crate::parser::Parser;

use crate::SyntaxKind::*;

impl<'a> Parser<'a> {
    pub(crate) fn parse_function(&mut self, checkpoint: rowan::Checkpoint) {
        self.start_node_at(checkpoint, FN_DEF);

        self.expect(T![fn]);

        self.ident();

        if self.current() == T![<] {
            self.parse_type_params(false);
        }

        if self.current() == T!["("] {
            self.parse_func_params(T!(")"));
        }

        if self.current() == T![->] {
            self.parse_return_type();
        }

        self.parse_block();

        self.finish_node()
    }

    pub(crate) fn parse_return_type(&mut self) {
        self.start_node(RET_TYPE);
        self.expect(T![->]);
        self.parse_type();
        self.finish_node()
    }
}

#[cfg(test)]
mod tests {
    test_parser! {parse_function,"fn main() {}"}
    test_parser! {parse_exported_function,"export fn main() {}"}
}
