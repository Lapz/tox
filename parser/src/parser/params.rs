use crate::parser::Parser;
use syntax::T;

use crate::SyntaxKind::{self, *};

impl<'a> Parser<'a> {
    pub(crate) fn parse_func_params(&mut self, flavour: SyntaxKind) {
        self.start_node(PARAM_LIST);

        // We assume that we have parsed the opening token

        self.bump();

        while !self.at(EOF) && !self.at(flavour) {
            self.func_param();

            if !self.at(flavour) && !self.expected(T![,]) {
                break;
            }
        }

        self.expect(flavour);

        self.finish_node()
    }

    pub(crate) fn func_param(&mut self) {
        self.start_node(PARAM);
        self.parse_pattern(false);

        if self.at(T![:]) {
            self.expect(T![:]);
            self.parse_type();
        }

        self.finish_node();
    }
}
