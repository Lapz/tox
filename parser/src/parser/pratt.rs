use crate::SyntaxKind::{self, *};
use std::any::Any;
use std::fmt::Debug;
use syntax::T;

use crate::parser::Parser;

pub trait Rule {
    fn rule(&self) -> RuleToken;
}
pub trait PrefixParser {
    fn parse(&self, parser: &mut Parser<'_>);
}
pub trait InfixParser: Debug + Any {
    fn parse(&self, parser: &mut Parser<'_>, checkpoint: rowan::Checkpoint);
    fn pred(&self) -> Precedence;
}

#[derive(Debug, Clone, Copy, PartialOrd, PartialEq)]
pub enum Precedence {
    None,
    Assignment,
    Or,
    And,
    Equality,
    Comparison,
    Term,
    Factor,
    Unary,
    Call,
    Primary,
}

#[derive(Debug, Eq, Hash, PartialEq, Clone, Copy)]
pub enum RuleToken {
    LParen,
    LBrace,
    Ident,
    Dot,
    LBracket,
    Minus,
    Plus,
    Slash,
    Star,
    Literal,
    Pipe,
    None,
    Excl,
    Comparison,
    Eq,
    EqEq,
    PlusEq,
    MinusEq,
    StarEq,
    SlashEq,
    AmpAmp,
    PipePipe,
    ColonColon,
}

impl Precedence {
    pub fn higher(self) -> Precedence {
        match self {
            Precedence::None | Precedence::Assignment => Precedence::Or,
            Precedence::Or => Precedence::And,
            Precedence::And => Precedence::Equality,
            Precedence::Equality => Precedence::Comparison,
            Precedence::Comparison => Precedence::Term,
            Precedence::Term => Precedence::Factor,
            Precedence::Factor => Precedence::Unary,
            Precedence::Unary => Precedence::Call,
            Precedence::Call => Precedence::Primary,
            Precedence::Primary => Precedence::Primary,
        }
    }
}

impl Rule for SyntaxKind {
    fn rule(&self) -> RuleToken {
        match self {
            INT_NUMBER | FLOAT_NUMBER | STRING | T![nil] | T![true] | T![false] => {
                RuleToken::Literal
            }
            IDENT => RuleToken::Ident,
            T![::] => RuleToken::ColonColon,
            T![=] => RuleToken::Eq,
            T![.] => RuleToken::Dot,
            T![+=] => RuleToken::PlusEq,
            T![-=] => RuleToken::MinusEq,
            T![/=] => RuleToken::SlashEq,
            T![*=] => RuleToken::MinusEq,
            T![+] => RuleToken::Plus,
            T![/] => RuleToken::Slash,
            T![!] => RuleToken::Excl,
            T![*] => RuleToken::Star,
            T![-] => RuleToken::Minus,
            T!["("] => RuleToken::LParen,
            T!["["] => RuleToken::LBracket,
            T!["{"] => RuleToken::LBrace,
            T![!=] | T![<] | T![>] | T![<=] | T![>=] => RuleToken::Comparison,
            T![&&] => RuleToken::AmpAmp,
            T![||] => RuleToken::PipePipe,
            T![|] => RuleToken::Pipe,
            T![==] => RuleToken::EqEq,
            _ => RuleToken::None,
        }
    }
}
