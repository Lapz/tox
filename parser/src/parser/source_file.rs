use crate::ast::*;
use crate::parser::Parser;
use rowan::GreenNodeBuilder;
use syntax::T;

use crate::{AstNode, SyntaxKind::*, SyntaxNode};

impl<'a> Parser<'a> {
    pub fn parse_program(&mut self) -> SourceFile {
        self.start_node(SOURCE_FILE);

        while !self.at(EOF) && !self.at(ERROR) {
            let has_visibility = self.has_visibility();
            let checkpoint = self.checkpoint();

            if has_visibility {
                self.parse_visibility();
            }

            match self.current() {
                T![type] => self.parse_type_alias(checkpoint),
                T![fn] => {
                    self.parse_function(checkpoint);
                }
                T![mod] => self.parse_module(),
                T![import] => self.parse_import(),
                T![enum] => self.parse_enum(checkpoint),
                T![class] => self.parse_class(checkpoint),
                T!["//"] => {
                    self.bump();
                    continue;
                }
                IDENT => {
                    let message = format!("Unknown top level modifier");

                    self.error(
                        message,
                        "Only `type`, `fn`, `mod`, `import`, `enum` , `class` are allowed at the top level",
                    )
                }

                _ => self.recover(),
            }
        }

        self.finish_node();

        let mut _builder = GreenNodeBuilder::new();

        std::mem::swap(&mut self.builder, &mut _builder);

        let green = _builder.finish();

        let root = SyntaxNode::new_root(green);

        SourceFile::cast(root).unwrap()
    }

    pub(crate) fn has_visibility(&mut self) -> bool {
        match self.current() {
            T![export] => true,
            _ => false,
        }
    }
}
