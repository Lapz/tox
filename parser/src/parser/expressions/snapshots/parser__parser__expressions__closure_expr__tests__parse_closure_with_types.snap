---
source: parser/src/parser/expressions/closure_expr.rs
expression: "crate::utils::dump_debug(&parser_output)"
---
SOURCE_FILE@[0; 83)
  FN_DEF@[0; 83)
    FN_KW@[0; 2) "fn"
    WHITESPACE@[2; 3) " "
    NAME@[3; 7)
      IDENT@[3; 7) "main"
    PARAM_LIST@[7; 9)
      L_PAREN@[7; 8) "("
      R_PAREN@[8; 9) ")"
    WHITESPACE@[9; 10) " "
    BLOCK_EXPR@[10; 83)
      BLOCK@[10; 83)
        L_CURLY@[10; 11) "{"
        WHITESPACE@[11; 20) "\n        "
        LET_STMT@[20; 82)
          LET_KW@[20; 23) "let"
          WHITESPACE@[23; 24) " "
          BIND_PAT@[24; 25)
            NAME@[24; 25)
              IDENT@[24; 25) "a"
          WHITESPACE@[25; 26) " "
          EQ@[26; 27) "="
          WHITESPACE@[27; 28) " "
          CLOSURE_EXPR@[28; 77)
            PARAM_LIST@[28; 33)
              PIPE@[28; 29) "|"
              PARAM@[29; 30)
                BIND_PAT@[29; 30)
                  NAME@[29; 30)
                    IDENT@[29; 30) "x"
              COMMA@[30; 31) ","
              PARAM@[31; 32)
                BIND_PAT@[31; 32)
                  NAME@[31; 32)
                    IDENT@[31; 32) "y"
              PIPE@[32; 33) "|"
            WHITESPACE@[33; 34) " "
            RET_TYPE@[34; 41)
              FRETURN@[34; 36) "->"
              WHITESPACE@[36; 37) " "
              IDENT_TYPE@[37; 41)
                IDENT@[37; 40) "i32"
                WHITESPACE@[40; 41) " "
            BLOCK_EXPR@[41; 77)
              BLOCK@[41; 77)
                L_CURLY@[41; 42) "{"
                WHITESPACE@[42; 55) "\n            "
                EXPR_STMT@[55; 65)
                  RETURN_EXPR@[55; 65)
                    RETURN_KW@[55; 61) "return"
                    WHITESPACE@[61; 62) " "
                    BIN_EXPR@[62; 65)
                      IDENT_EXPR@[62; 63)
                        NAME@[62; 63)
                          IDENT@[62; 63) "x"
                      PLUS@[63; 64) "+"
                      IDENT_EXPR@[64; 65)
                        NAME@[64; 65)
                          IDENT@[64; 65) "y"
                SEMI@[65; 66) ";"
                WHITESPACE@[66; 76) " \n        "
                R_CURLY@[76; 77) "}"
          WHITESPACE@[77; 82) "\n    "
        R_CURLY@[82; 83) "}"

